From e645477f7c9a5ca3c30902b1b89290896bbc0847 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Tue, 31 May 2016 20:23:24 +0100
Subject: [PATCH 4/4] VS2008: Add typedefs for more ?int??_t and define _STDINT

This allows us to detect later that it's already been done
as we need to in:
qtbase/src/3rdparty/harfbuzz-ng/src/hb-common.h
and:
qt3d/src/3rdparty/assimp/include/assimp/Compiler/pstdint.h

Also, pstdint.h defines _STDINT so if it got included first
then we don't define it here. What a mess.

Change-Id: Ic6b850f85fd620c2ae6232e35a230f9c1f9f8afd
---
 src/3rdparty/harfbuzz-ng/src/hb-common.h |  2 ++
 src/gui/opengl/qopenglext.h              | 42 +++++++++++++++++++++++++++++---
 2 files changed, 41 insertions(+), 3 deletions(-)

diff --git a/src/3rdparty/harfbuzz-ng/src/hb-common.h b/src/3rdparty/harfbuzz-ng/src/hb-common.h
index c291dbb..dc7879c 100644
--- qtbase/src/3rdparty/harfbuzz-ng/src/hb-common.h
+++ qtbase/src/3rdparty/harfbuzz-ng/src/hb-common.h
@@ -53,6 +53,7 @@
 #  include <sys/inttypes.h>
 /* VS 2010 (_MSC_VER 1600) has stdint.h */
 #elif defined (_MSC_VER) && _MSC_VER < 1600
+#if !defined(_STDINT)
 typedef __int8 int8_t;
 typedef unsigned __int8 uint8_t;
 typedef __int16 int16_t;
@@ -61,6 +62,7 @@ typedef __int32 int32_t;
 typedef unsigned __int32 uint32_t;
 typedef __int64 int64_t;
 typedef unsigned __int64 uint64_t;
+#endif
 #else
 #  include <stdint.h>
 #endif
diff --git a/src/gui/opengl/qopenglext.h b/src/gui/opengl/qopenglext.h
index 72316ca..e3d2f41 100644
--- qtbase/src/gui/opengl/qopenglext.h
+++ qtbase/src/gui/opengl/qopenglext.h
@@ -1402,9 +1402,45 @@ typedef unsigned long long int uint64_t;
 #elif defined(_WIN32) && (defined(__GNUC__) || (defined(_MSC_VER) && _MSC_VER >= 1600))
 #include <stdint.h>
 #elif defined(_WIN32)
-typedef __int32 int32_t;
-typedef __int64 int64_t;
-typedef unsigned __int64 uint64_t;
+/* Define _STDINT when we typedef these so we can know later not to
+ * do something daft like defining macros for them. Also define other
+ * things that may reasonably be expected to be defined by stdint.h.
+ */
+#if !defined(_STDINT)
+#define _STDINT
+typedef signed char int8_t;
+typedef short int16_t;
+typedef int int32_t;
+typedef long long int64_t;
+typedef unsigned char uint8_t;
+typedef unsigned short uint16_t;
+typedef unsigned int uint32_t;
+typedef unsigned long long uint64_t;
+#define INT8_C(x) (x)
+#define INT16_C(x) (x)
+#define INT32_C(x) (x)
+#define INT64_C(x) (x ## LL)
+#define UINT8_C(x) (x)
+#define UINT16_C(x) (x)
+#define UINT32_C(x) (x ## U)
+#define UINT64_C(x) (x ## ULL)
+#define INTMAX_C(x) INT64_C(x)
+#define UINTMAX_C(x) UINT64_C(x)
+#define INT8_MIN (-127i8 - 1)
+#define INT16_MIN (-32767i16 - 1)
+#define INT32_MIN (-2147483647i32 - 1)
+#define INT64_MIN (-9223372036854775807i64 - 1)
+#define INT8_MAX 127i8
+#define INT16_MAX 32767i16
+#define INT32_MAX 2147483647i32
+#define INT64_MAX 9223372036854775807i64
+#define UINT8_MAX 0xffui8
+#define UINT16_MAX 0xffffui16
+#define UINT32_MAX 0xffffffffui32
+#define UINT64_MAX 0xffffffffffffffffui64
+#define PRINTF_INT32_MODIFIER ""
+#define PRINTF_INT64_MODIFIER "ll"
+#endif
 #else
 /* Fallback if nothing above works */
 #include <inttypes.h>
-- 
2.8.2

